numpy/lib/nanfunctions.py

```diff
@@ -40,6 +40,28 @@
     ]
 
 
+def _nan_mask(a, out=None):
+    """
+    Parameters
+    ----------
+    a : array-like
+        Input array with at least 1 dimension.
+    out : ndarray, optional
+        Alternate output array in which to place the result.  The default
+        is ``None``; if provided, it must have the same shape as the
+        expected output and will prevent the allocation of a new array.
+
+    Returns
+    -------
+    y : ndarray
+        A bool array where ``np.nan`` positions are marked with ``False``
+        and other positions are marked with ``True``.
+    """
+    a = np.asarray(a)
```

Comment:
I think that for the private version we should be able to assume that `a` has already been turned into an array, no?

But more generally: we should short-circuit the cases where `nan` cannot exist, ie., something like
```
if a.dtype.kind not in 'fc':
    return True
```
For the later implementation, it will be important that this is `True` itself, since then the `where=True` argument will just be ignored.